MoltenVK/MoltenVK/Commands/MVKCmdDraw.mm

```diff
@@ -112,112 +112,99 @@
                 if (pipeline->needsVertexOutputBuffer()) {
                     vtxOutBuff = cmdEncoder->getTempMTLBuffer(_vertexCount * _instanceCount * 4 * cmdEncoder->_pDeviceProperties->limits.maxVertexOutputComponents);
-                    [cmdEncoder->_mtlRenderEncoder setVertexBuffer: vtxOutBuff->_mtlBuffer
-                                                            offset: vtxOutBuff->_offset
-                                                           atIndex: pipeline->getOutputBufferIndex().stages[kMVKShaderStageVertex]];
-                    // The shader only needs the number of vertices, so that's all we'll give it.
-                    // It'd be nice to be able to use setVertexBytes(), but since we can't guarantee
-                    // more than 4 bytes alignment because of indirect draws, we're stuck doing this.
-                    vtxParamsBuff = cmdEncoder->getTempMTLBuffer(4);
-                    *(uint32_t*)vtxParamsBuff->getContents() = _vertexCount;
-                    [cmdEncoder->_mtlRenderEncoder setVertexBuffer: vtxParamsBuff->_mtlBuffer
-                                                            offset: vtxParamsBuff->_offset
-                                                           atIndex: pipeline->getIndirectParamsIndex().stages[kMVKShaderStageVertex]];
-                }
-                if (cmdEncoder->_pDeviceMetalFeatures->baseVertexInstanceDrawing) {
-                    [cmdEncoder->_mtlRenderEncoder drawPrimitives: MTLPrimitiveTypePoint
-                                                      vertexStart: _firstVertex
-                                                      vertexCount: _vertexCount
-                                                    instanceCount: _instanceCount
-                                                     baseInstance: _firstInstance];
-                } else {
-                    [cmdEncoder->_mtlRenderEncoder drawPrimitives: MTLPrimitiveTypePoint
-                                                      vertexStart: _firstVertex
-                                                      vertexCount: _vertexCount
-                                                    instanceCount: _instanceCount];
+                    [mtlTessCtlEncoder setBuffer: vtxOutBuff->_mtlBuffer
+                                          offset: vtxOutBuff->_offset
+                                         atIndex: pipeline->getOutputBufferIndex().stages[kMVKShaderStageVertex]];
                 }
+				[mtlTessCtlEncoder setStageInRegion: MTLRegionMake2D(_firstVertex, _firstInstance, _vertexCount, _instanceCount)];
+				id<MTLComputePipelineState> vtxState = pipeline->getTessVertexStageState();
+				if (cmdEncoder->getDevice()->_pMetalFeatures->nonUniformThreadgroups) {
+					[mtlTessCtlEncoder dispatchThreads: MTLSizeMake(_vertexCount, _instanceCount, 1)
+                                 threadsPerThreadgroup: MTLSizeMake(vtxState.threadExecutionWidth, 1, 1)];
```

Comment:
To shut the compiler up for tvOS, can you wrap this in `#if MVK_MACOS_OR_IOS`, similar to what we did with `drawPatches:patchIndexBuffer:patchIndexBufferOffset:indirectBuffer:indirectBufferOffset:` in the same file?

There are 5 occurrences of this in this file. I won't bother marking them all.

Reply:
Thanks.

Sorry to be pedantic...but wouldn't wrapping `dispatchThreads:threadsPerThreadgroup:` by itself work the same, since `nonUniformThreadgroups = false` already guards the missing flow for tvOS?

If so, in doing so, we could get rid of the redundant calls to `dispatchThreadgroups:threadsPerThreadgroup:`.